Wrap names in __init__ as tuple if not already a tuple.

rpadaki opened the pull request:
This prevents undesired behavior when `names` is an iterable.
For example, it fixes the fact that `ntorch.ones(1,2, names="ab")` runs without error.
Meanwhile, it avoids the unhelpful error messages when running, say, `ntorch.ones(3, names=("vocab"))`, which without the fix results in `AssertionError: Tensor has  1 dim, but only 5 names`.

Fixed a typo'd extra space in the above AssertionError message.

rpadaki committed:
Wrap names in __init__ as tuple if not already a tuple.

Prevents undesired behavior when names is an iterable.

For example, fixes the fact that ntorch.ones(1,2, names="ab") runs without error.

srush commented:
This looks great, but let's be more aggressive. 

Can you put this function in a utils lib and call it? 

https://github.com/harvardnlp/namedtensor/blob/master/namedtensor/torch_helpers.py#L5

I think it should work if you pass a tuple or a list, but I agree not a string.

rpadaki committed:
Factored `make_tuple` to utils.py.

rpadaki committed:
Import `make_tuple` from utils.py in schema.py

rpadaki committed:
Removed `make_tuple` from `__init__` in core.py by factoring into `__init__` in schema.py.

Handle the case where `names` is an instance of `_Schema` in `build` method of schema.py before running `make_tuple`.

rpadaki committed:
Style fixes: Added double newline after imports

rpadaki commented:
I factored `make_tuple` out wherever I could find it.

The CI checks only passed with my first commit because I inadvertently set `names=(names,)` only after `names` was used in `__init__`.

The errors were specifically with the case handled in the `build` method on line 22 of [schema.py](https://github.com/harvardnlp/namedtensor/blob/master/namedtensor/schema.py#L23), which is for example initiated via the call on line 63 of [distributions.py](https://github.com/harvardnlp/namedtensor/blob/master/namedtensor/distributions.py#L63).

The fix I made was to let `if isinstance(names, _Schema)` take priority over `make_tuple` in the `build` method of utils.py. However, I'm not the biggest fan of this ad-hoc type-checking, and would prefer to either:
1. Rewrite the package and all existing code to make sure nobody ever passes a `_Schema` object as `names` (desirable if nobody would externally be passing schemas to ntorch.NamedTensor, undesirable if it significantly impairs efficiency)
2. Handle `if isinstance(names, _Schema)` in `make_tuple`, in utils.py (undesirable because utils.py should not depend on schema.py)

srush commented:
Cool, I'm going to accept for now because this is useful and then we can discuss more. 

I completely agree with your concerns. I don't think _Schema should be public. We should have a different way to build with schema.
